### PR TITLE
Support for the new builtin breakpoint function in Python 3.7

### PR DESCRIPTION
**Pull request by tonybaloney**
Working on conditions stated in #3180 

Support for Python 3.7s builtin breakpoint() method.

These are the behaviours:

- When breakpoint() is called and PYTHONBREAKPOINT is set to the default value, PyTest will use the Custom PDB trace UI instead of the system default Pdb.
- When tests are complete, the system will default back to the system Pdb trace UI.
- If --pdb is called on execution of Pytest, the custom Pdb interface is used on both breakpoint() and failed tests/unhandled exceptions.
- If --pdbcls is used, the custom class will be executed when a test fails (as expected within existing behaviour),  but also when breakpoint() is called from within a test, the custom class debugger will be instantiated.
- When breakpoint() is called and PYTHONBREAKPOINT is NOT set to the default value, PyTest will not get involved.

------------
Thanks for submitting a PR, your contribution is really appreciated!

Here's a quick checklist that should be present in PRs (you can delete this text from the final description, this is
just a guideline):

- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](/changelog/README.rst) for details.
- [ ] Target the `master` branch for bug fixes, documentation updates and trivial changes.
- [x] Target the `features` branch for new features and removals/deprecations.
- [x] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.

Unless your change is trivial or a small documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Add yourself to `AUTHORS` in alphabetical order;

### CODE REVIEW

**Comment by tonybaloney** [comment]
Discovered that the debugging module has no test coverage, so I want to develop tests for it before I start fiddling with it's behaviours.

**Comment by RonnyPfannschmidt** [comment]
@tonybaloney for historic reasons its in `testing/test_pdb.py`

**Comment by tonybaloney** [comment]
@RonnyPfannschmidt where would you recommend I put these tests? is `test_debugging` ok? Should I migrate the existing test case (perhaps in a seperate PR)?

**Comment by RonnyPfannschmidt** [comment]
@tonybaloney for now please put them into test-pdb, as a follow-up it seems sensible to rename that file

**Comment by tonybaloney** [comment]
This PR is ready for review, I still need to add some documentation to explain the behaviours, but would like feedback on the work so far (most of it was just tests tbh!)

**Comment by tonybaloney** [comment]
Tested with the example IPython class, which breaks but seems to be caused by https://github.com/pytest-dev/pytest/issues/2064 
```
(env) anthonyshaw ~/repo/test-pytest $ pytest test --pdbcls=IPython.terminal.debugger:TerminalPdb
=========================================================================================================== test session starts ============================================================================================================
platform darwin -- Python 3.7.0a4, pytest-3.4.3.dev70+g242fb785, py-1.5.3, pluggy-0.6.0
rootdir: /Users/anthonyshaw/repo/test-pytest, inifile:
collected 1 item

test/test_breakpoint.py
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
F                                                                                                                                                                                                            [100%]

================================================================================================================= FAILURES =================================================================================================================
_____________________________________________________________________________________________________________ test_breakpoint ______________________________________________________________________________________________________________

    def test_breakpoint():
>      breakpoint()

test/test_breakpoint.py:2:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../pytest/_pytest/debugging.py:78: in set_trace
    cls._pdb_cls().set_trace(frame)
env/lib/python3.7/site-packages/IPython/terminal/debugger.py:26: in __init__
    self.pt_init()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <IPython.terminal.debugger.TerminalPdb object at 0x10629ab70>

    def pt_init(self):
        def get_prompt_tokens(cli):
            return [(Token.Prompt, self.prompt)]

        def patch_stdout(**kwargs):
            return self.pt_cli.patch_stdout_context(**kwargs)

        if self._ptcomp is None:
            compl = IPCompleter(shell=self.shell,
                                        namespace={},
                                        global_namespace={},
                                        parent=self.shell,
                                       )
            self._ptcomp = IPythonPTCompleter(compl, patch_stdout=patch_stdout)

        kbmanager = KeyBindingManager.for_prompt()
        supports_suspend = Condition(lambda cli: hasattr(signal, 'SIGTSTP'))
        kbmanager.registry.add_binding(Keys.ControlZ, filter=supports_suspend
                                      )(suspend_to_bg)

        if self.shell.display_completions == 'readlinelike':
            kbmanager.registry.add_binding(Keys.ControlI,
                                 filter=(HasFocus(DEFAULT_BUFFER)
                                         & ~HasSelection()
                                         & ViInsertMode() | EmacsInsertMode()
                                         & ~cursor_in_leading_ws
                                         ))(display_completions_like_readline)
        multicolumn = (self.shell.display_completions == 'multicolumn')

        self._pt_app = create_prompt_application(
                            editing_mode=getattr(EditingMode, self.shell.editing_mode.upper()),
                            key_bindings_registry=kbmanager.registry,
                            history=self.shell.debugger_history,
                            completer= self._ptcomp,
                            enable_history_search=True,
                            mouse_support=self.shell.mouse_support,
                            get_prompt_tokens=get_prompt_tokens,
                            display_completions_in_columns=multicolumn,
>                           style=self.shell.style
        )
E       AttributeError: 'TerminalInteractiveShell' object has no attribute 'style'

env/lib/python3.7/site-packages/IPython/terminal/debugger.py:66: AttributeError
========================================================================================================= 1 failed in 0.29 seconds =========================================================================================================
```

**Comment by tonybaloney** [comment]
@nicoddemus @RonnyPfannschmidt done (except docs)

**Comment by tonybaloney** [comment]
Build failed on 1 test

https://travis-ci.org/pytest-dev/pytest/jobs/357232290#L585

I asserted the output was "1 passed" but it is "2 passed" on Travis. Might have something to do with Hypothesis, trying to repro locally, but I can make the assertion a bit more vauge

**Comment by tonybaloney** [comment]
Yep. once I install hypothesis, it changes to 2 passed locally as well (which I guess makes sense as the hypothesis plugin is designed to do that)

File: testing/test_pdb.py [hunk 1]
```diff
@@ -434,3 +463,115 @@ def test_foo():
...
+            assert a == 0
+        """)
+        result = testdir.runpytest_inprocess("", p1)
+        result.stdout.fnmatch_lines([
+            "* passed*",
+        ])
+        assert sys.breakpointhook != pytestPDB.set_trace
+
+    @pytest.mark.skipif(not SUPPORTS_BREAKPOINT_BUILTIN, reason="Requires breakpoint() builtin")
+    def test_sys_breakpointhook_configure_and_unconfigure_with_pdb_flag(self, testdir):
```
Comment (nicoddemus) [inline testing/test_pdb.py]: You can probably parametrize this test with the above.

```python
    @pytest.mark.skipif(not SUPPORTS_BREAKPOINT_BUILTIN, reason="Requires breakpoint() builtin")
    @pytest.mark.parametrize('args', [('--pdb',), ()])
    def test_sys_breakpointhook_configure_and_unconfigure(self, testdir, args):
        ...
        result = testdir.runpytest_inprocess(*args)
```

(It's not necessary to pass `p1` to `runpytest_inprocess`).

File: testing/test_pdb.py [hunk 2]
```diff
@@ -434,3 +463,115 @@ def test_foo():
...
+        rest = child.read().decode("utf8")
+        assert "1 failed" in rest
+        assert "reading from stdin while output" not in rest
+        TestPDB.flush(child)
+
+    @pytest.mark.skipif(not SUPPORTS_BREAKPOINT_BUILTIN, reason="Requires breakpoint() builtin")
+    def test_pdb_not_altered(self, testdir):
+        """
+        Test that calling PDB set_trace() is not affected
+        when PYTHONBREAKPOINT=0
```
Comment (nicoddemus) [inline testing/test_pdb.py]: Is this comment correct? I don't see `PYTHONBREAKPOINT` being set in the test..

File: testing/test_pdb.py [hunk 3]
```diff
@@ -1,11 +1,17 @@
...
 import platform
+import os
 
 import _pytest._code
+from _pytest.debugging import (SUPPORTS_BREAKPOINT_BUILTIN, pytestPDB,
+                               pytest_configure)
 import pytest
 
 
+_ENVIRON_PYTHONBREAKPOINT = getattr(os.environ, 'PYTHONBREAKPOINT', '')
```
Comment (nicoddemus) [inline testing/test_pdb.py]: Again this seems like it should be `_ENVIRON_PYTHONBREAKPOINT = os.environ.get('PYTHONBREAKPOINT', '')

File: testing/test_pdb.py [hunk 4]
```diff
@@ -434,3 +463,115 @@ def test_foo():
...
+
+    @pytest.mark.skipif(not SUPPORTS_BREAKPOINT_BUILTIN, reason="Requires breakpoint() builtin")
+    def test_sys_breakpointhook_configure_and_unconfigure(self, testdir):
+        """
+        Test that sys.breakpointhook is set to the custom Pdb class once configured, test that
+        hook is reset to system value once pytest has been unconfigured
+        """
+        config = testdir.parseconfig()
+
+        pytest_configure(config)
```
[the remark this answers is not in the document]
Comment (tonybaloney) [reply testing/test_pdb.py]: I couldn't locate the right test pattern looking at  the others. This makes sense

File: testing/test_pdb.py (hunk 3 above)
Comment (tonybaloney) [reply testing/test_pdb.py]: Fixed in f1f4c8c1

File: testing/test_pdb.py (hunk 2 above)
Comment (tonybaloney) [reply testing/test_pdb.py]: Removed in f1f4c8c1

File: testing/test_pdb.py (hunk 1 above)
Comment (tonybaloney) [reply testing/test_pdb.py]: Fixed in f1f4c8c1

File: testing/test_pdb.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tonybaloney) [reply testing/test_pdb.py]: Fixed in f1f4c8c1

File: testing/test_pdb.py [hunk 5]
```diff
@@ -33,6 +39,29 @@ def interaction(self, *args):
...
+
+        def interaction(self, *args):
+            called.append("interaction")
+
+        def set_trace(self, frame):
+            print("**CustomDebugger**")
+            called.append("set_trace")
+
+    _pytest._CustomDebugger = _CustomDebugger
+    return called
```
[the remark this answers is not in the document]
Comment (tonybaloney) [reply testing/test_pdb.py]: Done in f1f4c8c1

File: _pytest/debugging.py [hunk 6]
```diff
@@ -27,12 +34,20 @@ def pytest_configure(config):
     if config.getvalue("usepdb"):
         config.pluginmanager.register(PdbInvoke(), 'pdbinvoke')
 
+    # Use custom Pdb class set_trace instead of default Pdb on breakpoint() call
+    if SUPPORTS_BREAKPOINT_BUILTIN:
+        _environ_pythonbreakpoint = getattr(os.environ, 'PYTHONBREAKPOINT', '')
```
[the remark this answers is not in the document]
Comment (tonybaloney) [reply _pytest/debugging.py]: Fixed in f1f4c8c1

**Comment by tonybaloney** [comment]
@nicoddemus made the code changes. Docs still need improving, but explain the behaviour for now. I can add more examples later when I've got some more time

**Comment by nicoddemus** [comment]
Hi @tonybaloney thanks for the quick response! I made some doc formatting changes (all minor), hope you don't mind.

I worry though that the `os.environ` mistake was not caught by any tests. Can you design a test that fails with the incorrect code?

**Comment by tonybaloney** [comment]
@nicoddemus not sure I really understood this properly. It doesn't seem to execute. If I change the assertions it doesnt fail the tests 
https://github.com/pytest-dev/pytest/pull/3331/commits/f1f4c8c10481ab643a497876dfaed713f54c9479#diff-6244362351d7c228d95cde366267b10aR511

**Comment by nicoddemus** [comment]
@tonybaloney Oh my bad, my example was incomplete:

```python
...
testdir.makepyfile("""
    def test_nothing(): pass
""")
result = testdir.runpytest()
result.stdout.fnmatch_lines([
    '*1 passed*',
])
```

IOW, `testdir` lets you create test files, execute pytest and check the output. I forgot the last bit, sorry.

**Comment by tonybaloney** [comment]
@nicoddemus  that makes sense, but all 3 fail now 

```
  File "/Users/anthonyshaw/repo/pytest/env/lib/python3.7/site-packages/pluggy-0.6.0-py3.7.egg/pluggy/__init__.py", line 253, in register
    self._verify_hook(hook, hookimpl)
  File "/Users/anthonyshaw/repo/pytest/env/lib/python3.7/site-packages/pluggy-0.6.0-py3.7.egg/pluggy/__init__.py", line 361, in _verify_hook
    (hookimpl.plugin_name, hook.name))
pluggy.PluginValidationError: Plugin '/private/var/folders/5_/ljj1v1m91hl_2r_9t_00bhmc0000gn/T/pytest-of-anthonyshaw/pytest-67/test_sys_breakpointhook_configure_and_unconfigure0/conftest.py'
hook 'pytest_configure'
historic incompatible to hookwrapper
```

I looked at the documentation and https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_configure that hook doesn't support hookwrapper=True.

I tried taking that out and just get another error

```
>       self._match_lines(lines2, fnmatch, 'fnmatch')
E       Failed: nomatch: '*1 passed in *'
E           and: 'INTERNALERROR> Traceback (most recent call last):'
E           and: 'INTERNALERROR>   File "/Users/anthonyshaw/repo/pytest/_pytest/main.py", line 96, in wrap_session'
E           and: 'INTERNALERROR>     config._do_configure()'
E           and: 'INTERNALERROR>   File "/Users/anthonyshaw/repo/pytest/_pytest/config.py", line 922, in _do_configure'
E           and: 'INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))'
E           and: 'INTERNALERROR>   File "/Users/anthonyshaw/repo/pytest/env/lib/python3.7/site-packages/pluggy-0.6.0-py3.7.egg/pluggy/__init__.py", line 630, in call_historic'
E           and: 'INTERNALERROR>     proc(x)'
E           and: "INTERNALERROR> TypeError: 'NoneType' object is not callable"
E           and: ''
E       remains unmatched: '*1 passed in *'

/Users/anthonyshaw/repo/pytest/_pytest/pytester.py:1175: Failed
```

File: testing/test_pdb.py [hunk 7]
```diff
@@ -481,14 +480,15 @@ def test_supports_breakpoint_module_global(self):
...
 
     @pytest.mark.skipif(not SUPPORTS_BREAKPOINT_BUILTIN, reason="Requires breakpoint() builtin")
-    def test_sys_breakpointhook_not_custom_pdb(self):
+    def test_sys_breakpointhook_is_pytest_PDB(self):
         """
-        Test that sys.breakpointhook is not set to the custom Pdb class
+        Test that sys.breakpointhook by default is set to pytest's internal class.
         """
-        assert sys.breakpointhook != pytestPDB.set_trace
+        assert sys.breakpointhook == pytestPDB.set_trace
```
Comment (nicoddemus) [inline testing/test_pdb.py]: This test seemed incorrect to me and it was failing on my system: wasn't `sys.breakpoint` supposed to be set in the normal circumstance? I changed the logic and docstring to reflect that and it passes. @tonybaloney please double check here.

**Comment by nicoddemus** [comment]
@tonybaloney sorry my bad, I forgot `pytest_configure` cannot be hook-wrappers. I pushed some changes to fix the tests:

* Execute some of them in a separate process, so changes to global state doesn't affect the rest of the tests.
* I had to use the internal `config._cleanup` mechanism to ensure the hook is restored at the end of the session.
* Fixed the parametrization in two of the tests.
* I had to change the logic of one of the tests, see https://github.com/pytest-dev/pytest/pull/3331/commits/804392e5f2222fc2c83d7a8a098b58ca4c126195#r177918076.

**Comment by tonybaloney** [comment]
@nicoddemus I figured out why that last test was failing, because the test is being run inside a pytest session, the new behaviour would mean that sys.breakpointhook is going to be unpredictable depending on the version of pytest being used and it doesn't test anything other than runtime assumptions.
So I removed it.

I'm all done now :)

File: testing/test_pdb.py [hunk 8]
```diff
@@ -434,3 +463,122 @@ def test_foo():
...
+
+
+class TestDebuggingBreakpoints(object):
+
+    def test_supports_breakpoint_module_global(self):
+        """
+        Test that supports breakpoint global marks on Python 3.7+ and not on
+        CPython 3.5, 2.7
+        """
+        if sys.version_info.major == 3 and sys.version_info.minor >= 7:
```
Comment (asottile) [inline testing/test_pdb.py]: could use `sys.version_info >= (3, 7)` as a slightly more readable alternative.

File: AUTHORS [hunk 9]
```diff
@@ -17,6 +17,7 @@ Andreas Zeidler
 Andrzej Ostrowski
 Andy Freeland
 Anthon van der Neut
+Anthony Shaw
 Anthony Sottile
```
Comment (asottile) [inline AUTHORS]: well hello fellow anthony :D

**Comment by RonnyPfannschmidt** [comment]
well done, im happy to merge :+1:

**Pull request merged**